From a009b96b6a39744e6c02ea036f1719ae74935424 Mon Sep 17 00:00:00 2001
From: Sebastian Bazley <sebb@apache.org>
Date: Thu, 10 Mar 2011 16:36:48 +0000
Subject: [PATCH] NET-327 "Unconnected sockets not implemented" when using
 FTPSClient Added disconnect() override which resets the socket factories to
 their defaults

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/net/trunk@1080269 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                       |  4 +++
 .../apache/commons/net/ftp/FTPSClient.java    | 26 +++++++++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 9b8a314dc..a6535aa52 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -57,6 +57,10 @@ The <action> type attribute can be add,update,fix,remove.
 
     <body>
         <release version="3.0" date="TBA" description="TBA">
+            <action issue="NET-327" dev="sebb" type="add" due-to="Bogdan Drozdowski" due-to-email="bogdandr # op . pl">
+            "Unconnected sockets not implemented" when using FTPSClient
+            Added disconnect() override which resets the socket factories to their defaults
+            </action>
             <action issue="NET-350" dev="sebb" type="add" due-to="Bogdan Drozdowski" due-to-email="bogdandr # op . pl">
             "java.net.SocketException: Broken pipe" when calling "TelnetClient.sendAYT()"
             Added SocketClient#isAvailable() method to perform additional checks on a socket.
diff --git a/src/main/java/org/apache/commons/net/ftp/FTPSClient.java b/src/main/java/org/apache/commons/net/ftp/FTPSClient.java
index b39ab38fe..ee9e95a55 100644
--- a/src/main/java/org/apache/commons/net/ftp/FTPSClient.java
+++ b/src/main/java/org/apache/commons/net/ftp/FTPSClient.java
@@ -33,6 +33,8 @@
 import javax.net.ssl.SSLSocketFactory;
 import javax.net.ssl.TrustManager;
 
+import sun.awt.windows.ThemeReader;
+
 /**
  * FTP over SSL processing. If desired, the JVM property -Djavax.net.debug=all can be used to
  * see wire-level SSL details.
@@ -422,6 +424,11 @@ public void execPBSZ(long pbsz) throws SSLException, IOException {
      * S - Safe(SSL protocol only)</br>
      * E - Confidential(SSL protocol only)</br>
      * P - Private
+     * <p>
+     * <b>N.B.</b> the method calls
+     *  {@link #setSocketFactory(javax.net.SocketFactory)} and
+     *  {@link #setServerSocketFactory(javax.net.ServerSocketFactory)}
+     *  
      * @param prot Data Channel Protection Level.
      * @throws SSLException If the server reply code does not equal "200".
      * @throws IOException If an I/O error occurs while sending
@@ -541,4 +548,23 @@ public TrustManager getTrustManager() {
     public void setTrustManager(TrustManager trustManager) {
         this.trustManager = trustManager;
     }
+
+    /**
+     * Closes the connection to the FTP server and restores
+     * connection parameters to the default values.
+     * <p>
+     * Calls {@code setSocketFactory(null)} and {@code setServerSocketFactory(null)}
+     * to reset the factories that may have been changed during the session, 
+     * e.g. by {@link #execPROT(String)}
+     * @exception IOException If an error occurs while disconnecting.
+     * @since 3.0
+     */
+    @Override
+    public void disconnect() throws IOException
+    {
+        super.disconnect();
+        setSocketFactory(null);
+        setServerSocketFactory(null);
+    }
 }
+/* kate: indent-width 4; replace-tabs on; */
